org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/embedder/MavenExecutionContext.java

```diff
@@ -149,32 +152,59 @@ protected MavenExecutionRequest newExecutionRequest() throws CoreException {
       }
       request = DefaultMavenExecutionRequest.copy(parent);
     }
-    if(request == null) {
-      request = createExecutionRequest(IMavenConfiguration.getWorkspaceConfiguration(),
-          containerLookup, MavenPlugin.getMaven().getSettings());
-      request.setBaseDirectory(basedir);
+    if(request == null || !Objects.equals(request.getMultiModuleProjectDirectory(), multiModuleProjectDirectory)) {
+      IMavenConfiguration workspaceConfiguration = IMavenConfiguration.getWorkspaceConfiguration();
+      Optional<MavenProperties> mavenArgs;
+      try {
+        mavenArgs = MavenProperties.getMavenArgs(multiModuleProjectDirectory);
+      } catch(IOException ex) {
+        throw new CoreException(Status.error("Loading maven.config failed!", ex));
+      }
+      MavenSettingsLocations mavenSettingsLocations = mavenArgs
+          .map(mavenCfg -> mavenCfg.getSettingsLocations(workspaceConfiguration))
+          .orElseGet(workspaceConfiguration::getSettingsLocations);
+      if(request == null) {
+        //create a fresh one ....
+        request = createExecutionRequest(workspaceConfiguration, containerLookup, mavenSettingsLocations);
+        request.setBaseDirectory(basedir);
+      } else {
+        //update existing configuration, we might have copied from an outer context here but the multi module directory is different we need to update some things then...
+        Settings settings = MavenPlugin.getMaven().getSettings(mavenSettingsLocations);
+        File requestLocalRepositoryPath = request.getLocalRepositoryPath();
+        File settingsLocalRepositoryPath = getLocalRepositoryPath(settings);
+        if(!pathEquals(requestLocalRepositoryPath, settingsLocalRepositoryPath)) {
+          updateLocalRepository(request, settingsLocalRepositoryPath, containerLookup);
+        }
+        //TODO maybe also need to update user properties?!?
+        updateSettingsFiles(request, mavenSettingsLocations);
+      }
+      request.setMultiModuleProjectDirectory(multiModuleProjectDirectory);
     }
-    request.setMultiModuleProjectDirectory(multiModuleProjectDirectory);
-
     return request;
   }
 
+  private boolean pathEquals(File file1, File file2) {
+    if(file1 != null && file2 != null) {
```

Comment:
Could they both be null? That would make them equals

Reply:
Better safe than sorry, I adjusted the code accordingly.